GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m

```diff
@@ -211,9 +212,41 @@ - (instancetype)init {
   return self;
 }
 
+- (void)beginBackgroundTaskWithName:(NSString *)name
+                      estimatedTime:(NSTimeInterval)seconds
+                         usingBlock:(void (^)(void))block
+                  expirationHandler:(void (^)(void))handler {
+#if !TARGET_OS_WATCH
+  __block GDTCORBackgroundIdentifier bgID = GDTCORBackgroundIdentifierInvalid;
+  bgID = [self beginBackgroundTaskWithName:name
+                         expirationHandler:^{
+                           if (bgID != GDTCORBackgroundIdentifierInvalid) {
+                             if (handler) {
+                               handler();
+                             }
+                             [[GDTCORApplication sharedApplication] endBackgroundTask:bgID];
+                             bgID = GDTCORBackgroundIdentifierInvalid;
+                           }
+                         }];
+  block();
+  [self endBackgroundTask:bgID];
+  bgID = GDTCORBackgroundIdentifierInvalid;
+  return;
+#elif TARGET_OS_WATCH
+  [self performExpiringActivityWithReason:name
+                               usingBlock:^(BOOL expired) {
+                                 block();
+                                 if (expired && !seconds) {
+                                   [[NSRunLoop currentRunLoop]
+                                       runUntilDate:[NSDate dateWithTimeIntervalSinceNow:seconds]];
+                                 }
+                               }];
+#endif
+}
+
+#if !TARGET_OS_WATCH
 - (GDTCORBackgroundIdentifier)beginBackgroundTaskWithName:(NSString *)name
```

Comment:
Ok, so. I think I have an idea of how to keep this same API and have it function on watchOS. Essentially, we create a dispatch_semaphore and store it in a class-wide array mapping a generated ID (the next index) to the semaphore. We dispatch_wait in performExpiringActivity block, which should be called on a global queue so as not to block the current queue. If the block is called with expiring set to YES, we dispatch_signal the semaphore, effectively not waiting for the previous background task to finish. Otherwise, in endBackgroundTask, we find the semaphore we created, signal it, and that will effectively end the performExpiringActivity block. WDYT?